### ISSUE TITLE
Segfault from TabularEditor + `edit_traits`

### ISSUE THREAD

**Issue by scott-maddox** [enthought/traitsui#568]
I ran into an odd segfault and was able to reduce it to this minimal reproduction.

```python
from traits.api import (
    HasTraits, List, Unicode, Tuple
)
from traitsui.api import View, UItem, TabularEditor
from traitsui.tabular_adapter import TabularAdapter


class MyClass(HasTraits):
    my_list = List(Tuple(Unicode))
    selected_row = Tuple(Unicode)

    traits_view = View(
        UItem(
            'my_list',
            editor=TabularEditor(
                selected_row='selected_row',
                adapter=TabularAdapter(columns=['Features']),
            ),
        ),
    )


if __name__=='__main__':
    MyClass().edit_traits()

```

This segfaults for me on:

```
Enthought Deployment Manager -- https://www.enthought.com
Python 3.6.0 |Enthought, Inc. (x86_64)| (default, Mar  3 2017, 03:26:01) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import traits
>>> traits.__version__
'5.1.1'
>>> import traitsui
>>> traitsui.__version__
'6.1.1'
```

**Comment by rkern** [comment]
Note that `selected_row` should be an `Int` trait; it gets the _index_ of the selected row. `selected = WhateverActualValueTypeOfTheRowIs()` is [the trait to use](https://github.com/enthought/traitsui/blob/master/traitsui/editors/tabular_editor.py#L70-L72) if you want to deal with the actual values.

The segfault may be because the Qt bindings don't like an exception bubbling up to that particular event handler.

**Comment by scott-maddox** [comment]
`selected_row = Int` still segfaults

**Comment by corranwebster** [comment]
I can't reproduce this on current master - it is possible #588 fixed this.

**Comment by scott-maddox** [comment]
I'm still getting segfault on master.

**Comment by corranwebster** [comment]
Are you using PyQt4 or PyQt5?

**Comment by scott-maddox** [comment]
PyQt4

**Comment by corranwebster** [comment]
OK, I can replicate with PyQt 4.12.

However, I note that you are calling `edit_traits` rather than `configure_traits` and then immediately exiting without starting the event loop, which is not normal usage.

I don't see the issue if I use `configure_traits`, nor do I see it if I manually start up the `QApplication` after calling `edit_traits`.  I also don't see the issue with PyQt 5.7.

So my suspicion is that this is tickling some Qt or PyQt 4 bug, but it could also be that because the event loop is not being run, Qt is not getting a chance to finalize widget state and things are getting gc'd in random order by Python as it shuts down and we're getting dangling references at the C++ level as a result (similar to the problem that Mark is trying to diagnose).

I'm tempted to close this issue unless you are seeing it with `configure_traits` or after otherwise running the event loop, unless there is a reason you need this behaviour?

**Comment by scott-maddox** [comment]
No, it's only with edit_traits. I understand it's not proper usage, but segfault is not a particularly helpful response for newcomers who have not learned the proper usage. That's how this came up.
